GLSL inlined code

abestephensg opened the pull request:
This PR adds the ability to declare strings of GLSL code in a Halide expression and then call functions defined inside the strings as Extern type Halide Calls. The mechanism is useful to developers porting GLSL apps to Halide because it allows them to incrementally move functionality from GLSL into Halide code.

Support for generating extern calls in the GLSL backend is added, with new intrinsics for encapsulating GLSL source strings, as well as support for passing whole images through to extern calls in the backend. See [test/opengl/glsl_extern.cpp](https://github.com/halide/Halide/blob/glsl-inline-code/test/opengl/glsl_extern.cpp) for example usage.

dsharletg commented on src/CodeGen_Internal.cpp:
```diff
@@ -81,8 +81,18 @@ void Closure::visit(const Variable *op) {
-        vars[op->name] = op->type;
+        // Check if variable is a buffer, used outside of a load or store node
+        if (ends_with(op->name, ".buffer")) {
+            std::string name = op->name.substr(0,op->name.length()-7);
```
Minor nitpick: space after the comma in the arguments to substr

dsharletg commented on src/CodeGen_Internal.cpp:
```diff
@@ -81,8 +81,18 @@ void Closure::visit(const Variable *op) {
+        if (ends_with(op->name, ".buffer")) {
+            std::string name = op->name.substr(0,op->name.length()-7);
+            debug(3) << "Adding buffer " << name << " to closure\n";
+            BufferRef & ref = buffers[name];
```
Minor nitpick: remove space after BufferRef reference.

dsharletg commented:
BTW, this is a pretty cool change. I wanted something like this for OpenCL a while back, but did not implement it.

abadams commented on src/Lower.cpp:
```diff
@@ -808,7 +808,21 @@ class FindBuffers : public IRGraphVisitor {
     void visit(const Call *op) {
-        IRGraphVisitor::visit(op);
+
+        if (op->call_type == Call::Extern) {
```
Why is this code necessary? It looks like it'll skip over input images that are only ever used in an extern call. The Variable visitor below skips over variables with no associated Parameter object anyway.

abestephensg replied:
The variable ends up with an associated Parameter object:

In the test use case we are creating a variable named "i0" + ".buffer", in order to pass the image parameter i0 to an extern call.  Creating an ordinary variable with this name asserts in the test code expression, inside the CheckVars visitor of an assignment operator, because the variable "i0.buffer" is undefined (Function.cpp:95).

We can make the new variable instance a Parameter type variable (the rationale being that the image that we are attempting to access is a in fact parameter), but next we crash (or assert with the fix below) in Lower.cpp add_image_checks. The boxes_touched visitor ignores extern calls and there is no information about the bounds of this image.

This check in FindBuffers avoids tracking buffers in add_image_checks in the narrow case that they are only passed directly to extern calls.

abadams replied:
The original code was added because doing checks on buffers (even when only
passed through to extern stages) caught a particular error where it first
occurred, as opposed to many levels deep into a pipeline expressed with
lots of extern stages. That's a good thing that would be bad to lose.

Seems like in general these Variables would not be associated with a
parameter object (because it might be an internal pipeline stage). Is that
right? Hrm. Perhaps we need to sneak variables that end with .buffer
through CheckVars.


On Wed, Mar 11, 2015 at 11:23 AM, abestephensg <notifications@github.com>
wrote:

> In src/Lower.cpp
> <https://github.com/halide/Halide/pull/703#discussion_r26240467>:
>
> > @@ -808,7 +808,21 @@ class FindBuffers : public IRGraphVisitor {
> >      using IRGraphVisitor::visit;
> >
> >      void visit(const Call *op) {
> > -        IRGraphVisitor::visit(op);
> > +
> > +        if (op->call_type == Call::Extern) {
>
> The variable ends up with an associated Parameter object:
>
> In the test use case we are creating a variable named "i0" + ".buffer", in
> order to pass the image parameter i0 to an extern call. Creating an
> ordinary variable with this name asserts in the test code expression,
> inside the CheckVars visitor of an assignment operator, because the
> variable "i0.buffer" is undefined (Function.cpp:95).
>
> We can make the new variable instance a Parameter type variable (the
> rationale being that the image that we are attempting to access is a in
> fact parameter), but next we crash (or assert with the fix below) in
> Lower.cpp add_image_checks. The boxes_touched visitor ignores extern calls
> and there is no information about the bounds of this image.
>
> This check in FindBuffers avoids tracking buffers in add_image_checks in
> the narrow case that they are only passed directly to extern calls.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/703/files#r26240467>.
>

abadams commented on src/runtime/opengl.cpp:
```diff
@@ -821,8 +821,8 @@ WEAK int halide_opengl_device_malloc(void *user_context, buffer_t *buf) {
         global_state.BindTexture(GL_TEXTURE_2D, tex);
-        global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
-        global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+        global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
```
What's going on with this change? (Mostly curious)

abadams replied:
Actually it invalidates the comment (no interpolation)

abestephensg replied:
One of the use cases for this PR is to enable apps to use interpolated texture sampling by making "extern" calls to texture2D. This change enables this sampling on textures created by the Halide runtime. Halide ordinarily maps integer Func coordinates to texel center coordinates- so there is no change in behavior for the ordinary glsl_texture_load intrinsic case. I'll add this description to the comment.

abadams commented on test/opengl/glsl_extern.cpp:
```diff
@@ -0,0 +1,245 @@
+    }
+
+    // Using normalized texture coordinates via GLSL texture2D:
+    if (0) {
```
Are these supposed to be if(0)? Looks like debugging code.

abestephensg commented:
I've modified CheckVars to allow ".buffer" tagged variables through and removed the special case from FindBuffers. Since the variable no longer has a parameter, I added a case to QualifyExpr to avoid renaming the variable.

In the case that the only use of the image in the Func is when it is passed to the extern call, we must make sure that this variable is picked up by InferArguments. This visitor looks for variables that have either an associated Parameter or Buffer. I've switched the example in the test to use the Buffer.

PTAL

abadams commented:
Looks plausible, but all the extern stage tests fail for me.

dsharletg commented:
Going to close this, please re-open if it gets updated and we want to merge it.
